src/main/java/org/opensearch/knn/index/memory/SharedIndexStateManager.java

```diff
@@ -0,0 +1,151 @@
+            //
+            // Ref:
+            // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-
+            SharedIndexStateEntry entry = sharedIndexStateCache.computeIfAbsent(modelId, m -> {
+                log.info("Loading entry to shared index state cache for model {}", modelId);
+                long sharedIndexStateAddress = JNIService.initSharedIndexState(indexAddress, knnEngine);
+                return new SharedIndexStateEntry(new SharedIndexState(sharedIndexStateAddress, modelId, knnEngine));
+            });
+            entry.incRef();
+            return entry.getSharedIndexState();
+        } finally {
+            this.readWriteLock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Indicate that the {@link SharedIndexState} is no longer being used. If nothing else is using it, it will be
+     * removed from the cache and evicted.
+     *
+     * After calling this method, {@link SharedIndexState} should no longer be used by calling thread.
+     *
+     * @param sharedIndexState to return to the system.
+     */
+    public void release(SharedIndexState sharedIndexState) {
+        this.readWriteLock.writeLock().lock();
+
+        try {
+            if (!sharedIndexStateCache.containsKey(sharedIndexState.getModelId())) {
+                // This should not happen. Will log the error and return to prevent crash
+                log.error("Attempting to evict model from cache but it is not present: {}", sharedIndexState.getModelId());
+                this.readWriteLock.writeLock().unlock();
+                return;
+            }
+
+            long refCount = sharedIndexStateCache.get(sharedIndexState.getModelId()).decRef();
+            if (refCount <= 0) {
+                log.info("Evicting entry from shared index state cache for key {}", sharedIndexState.getModelId());
+                sharedIndexStateCache.remove(sharedIndexState.getModelId());
+                JNIService.freeSharedIndexState(sharedIndexState.getSharedIndexStateAddress(), sharedIndexState.getKnnEngine());
+            }
+        } finally {
+            this.readWriteLock.writeLock().unlock();
+        }
+    }
+
+    private static final class SharedIndexStateEntry {
+        @Getter
+        private final SharedIndexState sharedIndexState;
+        private final AtomicLong referenceCount;
+
+        /**
+         * Constructor
+         *
+         * @param sharedIndexState sharedIndexStateContext being wrapped
+         */
+        private SharedIndexStateEntry(SharedIndexState sharedIndexState) {
+            this.sharedIndexState = sharedIndexState;
+            this.referenceCount = new AtomicLong(0);
+        }
+
+        /**
+         * Increases reference count by 1
+         *
+         * @return ++referenceCount
+         */
+        private long incRef() {
+            return referenceCount.incrementAndGet();
```

Comment:
nit: may be void since you are not using incremented value?

Reply:
I went back and forth. It is required for decRef() so I figured for symmetry I would make it return long